Check remembered entities before remembering entity

jpeel opened the pull request:
Messages that come through for an entity before StartEntity
has been processed for that entity caused redundant persistence
of the entity.

jpeel committed:
Check remembered entities before remembering entity

Messages that come through for an entity before StartEntity
has been processed for that entity caused redundant persistence
of the entity.

jpeel commented:
Refs #25261

johanandren commented:
OK TO TEST

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -455,11 +455,14 @@ private[akka] trait RememberingShard { selfType: Shard ⇒
-        messageBuffers.append(id, msg, snd)
-        processChange(EntityStarted(id))(sendMsgBuffer)
+        if (state.entities.contains(id)) {
+          getEntity(id).tell(payload, snd)
```
Is there a risk that `messageBuffers` already holds some messages and this will bypass those and result in that messages are delivered in wrong order?

patriknw replied:
would it be more safe to 
```
if (state.entities.contains(id)) {
  messageBuffers.append(id, msg, snd)
  sendMsgBuffer(EntityStarted(id))
}
```

jpeel replied:
It isn't possible for messageBuffers to hold anything for this entity at this point in the code because deliverTo is only called from within deliverMessage when there isn't a message buffer for this id (see line 312).

patriknw replied:
ok, thanks, could you add a `require` that message buffer is empty here so that it doesn't go unnoticed if we change something here in the future

jpeel committed:
Add a require statement for future safety.

deliverTo expects that messageBuffers does not contain the entity
id when it is called, but that could change.

jpeel committed:
Add a message to the require statement.

jpeel committed:
Merge branch 'master' into wip-cluster-sharding-remember-entities-redundant-journals

jpeel commented:
Good idea on the require. I've added it.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -455,11 +455,15 @@ private[akka] trait RememberingShard { selfType: Shard ⇒
-        messageBuffers.append(id, msg, snd)
-        processChange(EntityStarted(id))(sendMsgBuffer)
+        if (state.entities.contains(id)) {
+          require(!messageBuffers.contains(id), "Message buffers contains id.")
```
would be nice to include the `id`
```
s"Message buffers contains id [$id]."
```

patriknw reviewed (approved):
LGTM, with small nitpick

jpeel committed:
Improve an error message in a require.

jpeel committed:
Merge branch 'wip-cluster-sharding-remember-entities-redundant-journals' of https://github.com/jpeel/akka into wip-cluster-sharding-remember-entities-redundant-journals

jpeel committed:
Merge branch 'master' into wip-cluster-sharding-remember-entities-redundant-journals

jpeel commented:
I made the nitpick change.

chbatey reviewed (approved):
LGTM
